File: go/core/extension/protocol.go
// Package extension is the Go engine's implementation of SEP, the Smooth
// Extension Protocol.
//
// An extension is a long-lived subprocess speaking JSON-RPC 2.0 over ndjson on
// its stdio (identical framing to MCP stdio). The canonical wire schemas live
// in the smooth-operator repo at spec/extension/; the types here are the Go
// host's view of that wire. Field names are snake_case to match the spec.
//
// This package is purely additive: nothing here runs unless a caller builds an
// ExtensionHost and registers its tools into core.AgentOptions.Tools. With no
// host built the agent loop behaves exactly as before.
//
// Layout mirrors the Rust reference (rust/smooth-operator-core/src/extension/):
//   - protocol.go  — JSON-RPC frames + typed method params/results.
//   - manifest.go  — extension.toml discovery, global+project merge, ${env:VAR}.
//   - process.go   — one subprocess: ndjson codec, pending map, generation guard.
//   - host.go      — ExtensionHost: hook chaining, event fanout, the delegate seam.
//   - tool_proxy.go — ExtensionTool: an extension tool as a core.Tool.
package extension

import (
	"encoding/json"
	"fmt"
)

// PROTOCOL_VERSION is the SEP protocol version this host implements.
const ProtocolVersion = 1

// JSON-RPC error codes. Standard range plus the SEP extensions documented in
// spec/extension/envelope.md.
const (
	CodeParseError     = -32700
	CodeInvalidRequest = -32600
	CodeMethodNotFound = -32601
	CodeInvalidParams  = -32602
	CodeInternalError  = -32603

	// CodeBlocked — a hook or policy vetoed the operation.
	CodeBlocked = -32000
	// CodeNoUI — ui/request in a headless/uncapable frontend.
	CodeNoUI = -32001
	// CodeNotTrusted — extension acted beyond its granted trust.
	CodeNotTrusted = -32002
	// CodeContextViolation — command-tier action attempted from an event-tier context.
	CodeContextViolation = -32003
	// CodeCapabilityDisabled — method requires a capability the handshake did not enable.
	CodeCapabilityDisabled = -32004
	// CodeCancelled — request cancelled via $/cancel.
	CodeCancelled = -32800
)

// SEP method names, centralized so the host and tests never spell one wrong.
const (
	MethodInitialize             = "initialize"
	MethodShutdown               = "shutdown"
	MethodPing                   = "ping"
	MethodEvent                  = "event"
	MethodHook                   = "hook"
	MethodToolExecute            = "tool/execute"
	MethodToolUpdate             = "tool/update"
	MethodCommandExecute         = "command/execute"
	MethodCommandComplete        = "command/complete"
	MethodCancel                 = "$/cancel"
	MethodRegistryUpdate         = "registry/update"
	MethodToolsSetActive         = "tools/set_active"
	MethodExecRun                = "exec/run"
	MethodUIRequest              = "ui/request"
	MethodLog                    = "log"
	MethodBusPublish             = "bus/publish"
	MethodSessionSendMessage     = "session/send_message"
	MethodSessionSendUserMessage = "session/send_user_message"
	MethodSessionAppendEntry     = "session/append_entry"
)

// RpcError is a JSON-RPC error object. It implements error.
type RpcError struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

// NewRpcError builds an RpcError with no data payload.
func NewRpcError(code int, message string) *RpcError {
	return &RpcError{Code: code, Message: message}
}

func (e *RpcError) Error() string {
	return fmt.Sprintf("JSON-RPC error %d: %s", e.Code, e.Message)
}

// Message is the JSON-RPC 2.0 envelope. All four frame shapes share this struct;
// which fields are present determines the shape:
//
//   - request: id + method (+ optional params)
//   - notification: method, no id
//   - success response: id + result
//   - error response: id + error
//
// omitempty keeps absent fields off the wire so a request never carries a
// result/error key. ID/Params/Result are json.RawMessage so an arbitrary
// inbound line round-trips (the id may be an int or a string).
type Message struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *RpcError       `json:"error,omitempty"`
}

// NewRequest builds a request frame.
func NewRequest(id json.RawMessage, method string, params json.RawMessage) Message {
	return Message{JSONRPC: "2.0", ID: id, Method: method, Params: params}
}

// NewNotification builds a notification frame (no id, no reply expected).
func NewNotification(method string, params json.RawMessage) Message {
	return Message{JSONRPC: "2.0", Method: method, Params: params}
}

// NewSuccess builds a success response frame echoing id.
func NewSuccess(id json.RawMessage, result json.RawMessage) Message {
	if len(result) == 0 {
		result = json.RawMessage("{}")
	}
	return Message{JSONRPC: "2.0", ID: id, Result: result}
}

// NewErrorResponse builds an error response frame echoing id (id may be nil).
func NewErrorResponse(id json.RawMessage, err *RpcError) Message {
	return Message{JSONRPC: "2.0", ID: id, Error: err}
}

func (m *Message) hasID() bool { return len(m.ID) > 0 && string(m.ID) != "null" }

// IsRequest reports whether this frame is a request (has both id and method).
func (m *Message) IsRequest() bool { return m.hasID() && m.Method != "" }

// IsNotification reports whether this frame is a notification (method, no id).
func (m *Message) IsNotification() bool { return !m.hasID() && m.Method != "" }

// IsResponse reports whether this frame is a response (has id, no method).
func (m *Message) IsResponse() bool { return m.hasID() && m.Method == "" }

// Tier is whether a dispatch may only observe (event) or may mutate the session
// (command). Session-mutating ext→host actions require command.
type Tier string

const (
	TierEvent   Tier = "event"
	TierCommand Tier = "command"
)

// Context is the dispatch context carried by every host→ext event/hook/tool/command.
type Context struct {
	Token string `json:"token"`
	Tier  Tier   `json:"tier"`
}

// HostInfo identifies the engine host to the extension at handshake.
type HostInfo struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

// WorkspaceInfo describes the workspace root and whether it is trusted.
type WorkspaceInfo struct {
	Root    string `json:"root"`
	Trusted bool   `json:"trusted"`
}

// SessionInfo optionally identifies the active session at handshake.
type SessionInfo struct {
	ID string `json:"id,omitempty"`
}

// InitializeParams is the host→ext handshake.
type InitializeParams struct {
	ProtocolVersion int                        `json:"protocol_version"`
	Host            HostInfo                   `json:"host"`
	Workspace       WorkspaceInfo              `json:"workspace"`
	Session         *SessionInfo               `json:"session,omitempty"`
	Mode            string                     `json:"mode"`
	UICapabilities  []string                   `json:"ui_capabilities,omitempty"`
	Flags           map[string]json.RawMessage `json:"flags,omitempty"`
	// CapabilitiesEnabled is an opaque capability map the host may advertise.
	CapabilitiesEnabled json.RawMessage `json:"capabilities_enabled,omitempty"`
}

// Validate enforces the required fields the JSON Schema marks required and that
// serde rejects in the Rust host: protocol_version must be present (>= 1).
func (p InitializeParams) Validate() error {
	if p.ProtocolVersion < 1 {
		return fmt.Errorf("initialize params: protocol_version is required and must be >= 1")
	}
	return nil
}

// ExtensionInfo identifies the extension in its handshake reply.
type ExtensionInfo struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

// ToolRegistration declares a tool the extension exposes.
type ToolRegistration struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Parameters  json.RawMessage `json:"parameters"`
	Deferred    bool            `json:"deferred,omitempty"`
}

// CommandRegistration declares a slash-command the extension exposes.
type CommandRegistration struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}

// ShortcutRegistration binds a keyboard chord to one of the extension's commands.
// Only frontends with a key surface (the TUI) honor these.
type ShortcutRegistration struct {
	Key         string `json:"key"`
	Command     string `json:"command"`
	Description string `json:"description,omitempty"`
}

// Registrations is the set of things an extension registers at handshake (or
// later via registry/update).
type Registrations struct {
	Tools         []ToolRegistration     `json:"tools,omitempty"`
	Commands      []CommandRegistration  `json:"commands,omitempty"`
	Flags         []string               `json:"flags,omitempty"`
	Shortcuts     []ShortcutRegistration `json:"shortcuts,omitempty"`
	Subscriptions []string               `json:"subscriptions,omitempty"`
}

// InitializeResult is the extension's handshake reply.
type InitializeResult struct {
	ProtocolVersion int           `json:"protocol_version"`
	Extension       ExtensionInfo `json:"extension"`
	Registrations   Registrations `json:"registrations"`
}

// HookParams is the host→ext hook intercept.
type HookParams struct {
	Hook    string          `json:"hook"`
	Context Context         `json:"context"`
	Input   json.RawMessage `json:"input"`
}

// HookAction is the discriminator of a HookOutcome.
type HookAction string

const (
	ActionContinue HookAction = "continue"
	ActionBlock    HookAction = "block"
	ActionModify   HookAction = "modify"
)

// HookOutcome is an extension's reply to a hook. It marshals tagged by action:
//
//	{ "action": "continue" }
//	{ "action": "block", "reason": "..." }
//	{ "action": "modify", "patch": { ... } }
type HookOutcome struct {
	Action HookAction
	// Reason is set only for a block outcome (optional on the wire).
	Reason string
	// Patch is set only for a modify outcome (required for modify).
	Patch json.RawMessage
}

// UnmarshalJSON enforces the tagged-union constraints the Rust serde enum does:
// an unknown action is rejected, and a modify without a patch is rejected.
func (h *HookOutcome) UnmarshalJSON(b []byte) error {
	var raw struct {
		Action string          `json:"action"`
		Reason *string         `json:"reason"`
		Patch  json.RawMessage `json:"patch"`
	}
	if err := json.Unmarshal(b, &raw); err != nil {
		return err
	}
	switch HookAction(raw.Action) {
	case ActionContinue:
		*h = HookOutcome{Action: ActionContinue}
	case ActionBlock:
		h.Action = ActionBlock
		h.Reason = ""
		h.Patch = nil
		if raw.Reason != nil {
			h.Reason = *raw.Reason
		}
	case ActionModify:
		if len(raw.Patch) == 0 {
			return fmt.Errorf("modify hook outcome missing patch")
		}
		h.Action = ActionModify
		h.Reason = ""
		h.Patch = raw.Patch
	default:
		return fmt.Errorf("unknown hook action %q", raw.Action)
	}
	return nil
}

// MarshalJSON serializes a HookOutcome tagged by action.
func (h HookOutcome) MarshalJSON() ([]byte, error) {
	switch h.Action {
	case ActionContinue:
		return []byte(`{"action":"continue"}`), nil
	case ActionBlock:
		out := map[string]any{"action": "block"}
		if h.Reason != "" {
			out["reason"] = h.Reason
		}
		return json.Marshal(out)
	case ActionModify:
		patch := h.Patch
		if len(patch) == 0 {
			patch = json.RawMessage("null")
		}
		return json.Marshal(map[string]any{"action": "modify", "patch": patch})
	default:
		return nil, fmt.Errorf("cannot marshal hook outcome with action %q", h.Action)
	}
}

// ToolExecuteParams is the host→ext dispatch of a registered tool.
type ToolExecuteParams struct {
	CallID    string          `json:"call_id"`
	Tool      string          `json:"tool"`
	Arguments json.RawMessage `json:"arguments"`
	Context   Context         `json:"context"`
}

// Validate enforces the required call_id the Rust host requires (serde rejects
// a missing call_id).
func (p ToolExecuteParams) Validate() error {
	if p.CallID == "" {
		return fmt.Errorf("tool/execute params: call_id is required")
	}
	return nil
}

// ToolExecuteResult is the ext→host reply to tool/execute.
type ToolExecuteResult struct {
	Content string          `json:"content"`
	IsError bool            `json:"is_error,omitempty"`
	Details json.RawMessage `json:"details,omitempty"`
}

// ToolUpdateParams is a progress notification for an in-flight tool/execute.
type ToolUpdateParams struct {
	CallID   string          `json:"call_id"`
	Message  string          `json:"message,omitempty"`
	Progress *float64        `json:"progress,omitempty"`
	Details  json.RawMessage `json:"details,omitempty"`
}

// EventParams is a fire-and-forget lifecycle/turn event delivered to a
// subscribed extension.
type EventParams struct {
	Event string `json:"event"`
	// Seq is the per-connection monotonic sequence. Absent (a nil pointer) on the
	// out-of-band events_lost marker.
	Seq     *uint64         `json:"seq,omitempty"`
	Context Context         `json:"context"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

// LogParams is a structured log line folded into host tracing.
type LogParams struct {
	Level   string          `json:"level"`
	Message string          `json:"message"`
	Fields  json.RawMessage `json:"fields,omitempty"`
}

// CommandExecuteParams is the host→ext dispatch of a registered slash-command.
type CommandExecuteParams struct {
	Command   string          `json:"command"`
	Context   Context         `json:"context"`
	Arguments json.RawMessage `json:"arguments,omitempty"`
}

// CommandExecuteResult is the ext→host reply to command/execute.
type CommandExecuteResult struct {
	Content string `json:"content,omitempty"`
}

// CommandCompleteParams asks an extension for argument completions.
type CommandCompleteParams struct {
	Command string  `json:"command"`
	Context Context `json:"context"`
	Partial string  `json:"partial,omitempty"`
}

// Completion is one argument-completion suggestion.
type Completion struct {
	Value       string `json:"value"`
	Description string `json:"description,omitempty"`
}

// CommandCompleteResult is the ext→host reply to command/complete.
type CommandCompleteResult struct {
	Completions []Completion `json:"completions,omitempty"`
}

File: typescript/core/src/extension/protocol.ts
/**
 * SEP wire protocol — JSON-RPC 2.0 frames and typed method params/results.
 *
 * SEP (the Smooth Extension Protocol) is JSON-RPC 2.0 over ndjson on an extension
 * subprocess's stdio. The canonical schemas live in the `smooth-operator` repo at
 * `spec/extension/`; the types here are the TS engine host's view of that wire —
 * the sibling of the Rust host's `extension/protocol.rs`. Field names are
 * `snake_case` because they ARE the wire.
 *
 * The author-side `@smooai/smooth-extension-sdk` ships its own copy of these
 * shapes; the host keeps them local rather than importing the SDK, which would
 * invert the dependency (the engine is the lower layer the SDK's host consumers
 * build on).
 */

/** The SEP protocol version this host implements. Effective = min(host, ext). */
export const PROTOCOL_VERSION = 1;

/** SEP method names, centralized so the host and tests never spell one wrong. */
export const method = {
    INITIALIZE: 'initialize',
    SHUTDOWN: 'shutdown',
    PING: 'ping',
    EVENT: 'event',
    HOOK: 'hook',
    TOOL_EXECUTE: 'tool/execute',
    TOOL_UPDATE: 'tool/update',
    COMMAND_EXECUTE: 'command/execute',
    COMMAND_COMPLETE: 'command/complete',
    CANCEL: '$/cancel',
    REGISTRY_UPDATE: 'registry/update',
    TOOLS_SET_ACTIVE: 'tools/set_active',
    EXEC_RUN: 'exec/run',
    UI_REQUEST: 'ui/request',
    LOG: 'log',
    BUS_PUBLISH: 'bus/publish',
    SESSION_SEND_MESSAGE: 'session/send_message',
    SESSION_SEND_USER_MESSAGE: 'session/send_user_message',
    SESSION_APPEND_ENTRY: 'session/append_entry',
} as const;

/**
 * JSON-RPC + SEP error codes (see `spec/extension/envelope.md`). Standard range
 * plus the SEP extensions.
 */
export const codes = {
    ParseError: -32700,
    InvalidRequest: -32600,
    MethodNotFound: -32601,
    InvalidParams: -32602,
    InternalError: -32603,
    /** A hook or policy vetoed the operation. */
    Blocked: -32000,
    /** `ui/request` in a headless/uncapable frontend. */
    NoUI: -32001,
    /** Extension acted beyond its granted trust. */
    NotTrusted: -32002,
    /** Command-tier action attempted from an event-tier context. */
    ContextViolation: -32003,
    /** Method requires a capability the handshake did not enable. */
    CapabilityDisabled: -32004,
    /** Request cancelled via `$/cancel`. */
    Cancelled: -32800,
} as const;

/** A JSON-RPC error object. */
export interface RpcErrorObject {
    code: number;
    message: string;
    data?: unknown;
}

/** An error carrying a JSON-RPC error code, thrown/returned for a remote error. */
export class RpcError extends Error {
    constructor(
        public readonly code: number,
        message: string,
        public readonly data?: unknown,
    ) {
        super(message);
        this.name = 'RpcError';
    }

    toObject(): RpcErrorObject {
        return this.data === undefined ? { code: this.code, message: this.message } : { code: this.code, message: this.message, data: this.data };
    }
}

/** A JSON-RPC id: an integer or a string (null only on a parse-error response). */
export type Id = number | string | null;

/**
 * The JSON-RPC 2.0 envelope. All four frame shapes share this type; which fields
 * are present determines the shape:
 * - request: `id` + `method` (+ optional `params`)
 * - notification: `method`, no `id`
 * - success response: `id` + `result`
 * - error response: `id` + `error`
 */
export interface Message {
    jsonrpc: '2.0';
    id?: Id;
    method?: string;
    params?: unknown;
    result?: unknown;
    error?: RpcErrorObject;
}

/** Build a request frame. */
export function request(id: Exclude<Id, null>, methodName: string, params: unknown): Message {
    return { jsonrpc: '2.0', id, method: methodName, params };
}

/** Build a notification frame (no id, no reply expected). */
export function notification(methodName: string, params: unknown): Message {
    return { jsonrpc: '2.0', method: methodName, params };
}

/** Build a success response frame echoing `id`. */
export function success(id: Id, result: unknown): Message {
    return { jsonrpc: '2.0', id, result };
}

/** Build an error response frame echoing `id`. */
export function errorResponse(id: Id, error: RpcErrorObject): Message {
    return { jsonrpc: '2.0', id, error };
}

/** True when this frame is a request (has both `id` and `method`). */
export function isRequest(m: Message): boolean {
    return m.id !== undefined && m.id !== null && m.method !== undefined;
}

/** True when this frame is a notification (has `method`, no `id`). */
export function isNotification(m: Message): boolean {
    return (m.id === undefined || m.id === null) && m.method !== undefined;
}

/** True when this frame is a response (has `id`, no `method`). */
export function isResponse(m: Message): boolean {
    return m.method === undefined && m.id !== undefined && m.id !== null;
}

// ---------------------------------------------------------------------------
// The two-tier dispatch context.
// ---------------------------------------------------------------------------

/**
 * Whether a dispatch may only observe (`event`) or may mutate the session
 * (`command`). Session-mutating ext→host actions require `command`.
 */
export type Tier = 'event' | 'command';

/** The dispatch context carried by every host→ext event/hook/tool/command. */
export interface Context {
    token: string;
    tier: Tier;
}

// ---------------------------------------------------------------------------
// initialize
// ---------------------------------------------------------------------------

export interface HostInfo {
    name: string;
    version: string;
}

export interface WorkspaceInfo {
    root: string;
    trusted: boolean;
}

export interface InitializeParams {
    protocol_version: number;
    host: HostInfo;
    workspace: WorkspaceInfo;
    session?: { id?: string };
    mode: string;
    ui_capabilities?: string[];
    /** Parsed values for the flags the extension declares (name → value). */
    flags?: Record<string, unknown>;
    capabilities_enabled?: Record<string, unknown>;
}

export interface ToolRegistration {
    name: string;
    description: string;
    /** JSON Schema for the tool's arguments. */
    parameters: Record<string, unknown>;
    deferred?: boolean;
}

export interface CommandRegistration {
    name: string;
    description: string;
}

/** A keyboard shortcut an extension binds to one of its commands. */
export interface ShortcutRegistration {
    /** A human-typed chord, e.g. `ctrl+p`; the frontend parses it. */
    key: string;
    /** The registered command this chord invokes (no leading `/`). */
    command: string;
    description?: string;
}

export interface Registrations {
    tools?: ToolRegistration[];
    commands?: CommandRegistration[];
    flags?: string[];
    shortcuts?: ShortcutRegistration[];
    subscriptions?: string[];
}

export interface InitializeResult {
    protocol_version: number;
    extension: { name: string; version: string };
    registrations?: Registrations;
}

// ---------------------------------------------------------------------------
// hook
// ---------------------------------------------------------------------------

export interface HookParams {
    hook: string;
    context: Context;
    input: unknown;
}

/** The extension's reply to a `hook`, tagged by `action`. */
export type HookOutcome =
    | { action: 'continue' }
    | { action: 'block'; reason?: string }
    | { action: 'modify'; patch: unknown };

/**
 * Parse an untyped `hook` reply into a {@link HookOutcome}, or throw when it is
 * malformed. Mirrors the Rust host's serde-tagged decode: a `modify` without a
 * `patch`, or an unknown `action`, is rejected (the host then treats it as a
 * failed hook step).
 */
export function parseHookOutcome(value: unknown): HookOutcome {
    if (typeof value !== 'object' || value === null) throw new RpcError(codes.InvalidParams, 'hook outcome is not an object');
    const action = (value as { action?: unknown }).action;
    if (action === 'continue') return { action: 'continue' };
    if (action === 'block') {
        const reason = (value as { reason?: unknown }).reason;
        return typeof reason === 'string' ? { action: 'block', reason } : { action: 'block' };
    }
    if (action === 'modify') {
        if (!('patch' in (value as object))) throw new RpcError(codes.InvalidParams, "hook 'modify' outcome missing patch");
        return { action: 'modify', patch: (value as { patch: unknown }).patch };
    }
    throw new RpcError(codes.InvalidParams, `unknown hook action: ${String(action)}`);
}

// ---------------------------------------------------------------------------
// tool/execute + tool/update
// ---------------------------------------------------------------------------

export interface ToolExecuteParams {
    call_id: string;
    tool: string;
    arguments: unknown;
    context: Context;
}

export interface ToolExecuteResult {
    content: string;
    is_error?: boolean;
    details?: unknown;
}

export interface ToolUpdateParams {
    call_id: string;
    message?: string;
    progress?: number;
    details?: unknown;
}

// ---------------------------------------------------------------------------
// event
// ---------------------------------------------------------------------------

export interface EventParams {
    event: string;
    /** Per-connection monotonic sequence; absent on the `events_lost` marker. */
    seq?: number;
    context: Context;
    payload?: unknown;
}

// ---------------------------------------------------------------------------
// command/execute + command/complete (host→ext)
// ---------------------------------------------------------------------------

export interface CommandExecuteParams {
    command: string;
    context: Context;
    arguments?: unknown;
}

export interface CommandExecuteResult {
    content?: string;
}

export interface CommandCompleteParams {
    command: string;
    context: Context;
    partial?: string;
}

export interface Completion {
    value: string;
    description?: string;
}

export interface CommandCompleteResult {
    completions: Completion[];
}

// ---------------------------------------------------------------------------
// session/* (ext→host) — all require COMMAND tier
// ---------------------------------------------------------------------------

/** How a `session/send_user_message` is delivered relative to the current turn. */
export type DeliverAs = 'steer' | 'follow_up' | 'next_turn';

export interface SessionSendMessageParams {
    context: Context;
    text: string;
    role?: 'user' | 'assistant';
}

export interface SessionSendUserMessageParams {
    context: Context;
    text: string;
    deliver_as?: DeliverAs;
}

export interface SessionAppendEntryParams {
    context: Context;
    entry: unknown;
}
